### ISSUE TITLE
bug: `forge flatten` should ignore intermediary files for imports

### ISSUE THREAD

**Issue by kopy-kat** [foundry-rs/foundry#7239]
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

0.2.0

### What command(s) is the bug in?

forge flatten

### Operating System

macOS (Apple Silicon)

### Describe the bug

In more complex codebases, you might want to import objects from files other than the file in which an object is originally defined, in order to make the codebase easier to read. A simplified example:

File `A`:
```solidity
import {Test} from "./Intermediary";
```

File `Intermediary`:
```solidity
import {Test} from "./Test";
import {Test2} from "./Test2";
```

File `Test`:
```solidity
contract Test {}
```

However, when running `forge flatten` (and when using `--verify`), all files are treated equally, even when they are only used as "intermediary files". This means that the flattened version of file A will also include `Test2`, which is never imported into A.

#### Proposed solution
The flattener should determine whether a file is used only as an intermediary and ignore it if it is, jumping to the file from which the used import is exported from.

**Comment by mattsse** [comment]
isn't this solveable by not importing `import {Test2} from "./Test2";` in `Intermediary`?

**Comment by kopy-kat** [comment]
No, the point is to have intermediary files from which you can selectively import in order to make the code more readable. 

One analogy is `index.js` files in js which can remap different exported functions or objects from a folder. For example, instead of the import path being `package/api/component/function.js` you can create `package/index.js` which declares exports to be made and then import from `package`.

I appreciate that this is not very common in solidity codebases today but makes code much more readable imo.

**Comment by mattsse** [comment]
ah I think I get it now,

the important part is this

```
import {Test} from "./Intermediary";
```

not sure if this can be done easily/safely because I think this would require resolving reexports like detecting that Test is not used from Intermediary, but is reexported in Intermediary 

wdyt @klkvr

**Comment by kopy-kat** [comment]
But `Test` cant be used in `Intermediary` right? Otherwise the compiler should throw an `Identifier already declared` error. 

As long as named imports are used I think this should be relatively straightforward, but if a file has multiple unnamed imports then it could be complicated to figure out where a specific object is coming from.

Another edge case would be aliased imports but this should also be relatively straightforward to handle

**Comment by kopy-kat** [comment]
hey guys, any update on this? I also believe this is the same as https://github.com/foundry-rs/foundry/issues/2266
